File: packages/google/openiap/src/main/java/dev/hyo/openiap/OpenIapError.kt
package dev.hyo.openiap

/**
 * OpenIAP specific exceptions
 */
sealed class OpenIapError : Exception() {
    abstract val code: String
    abstract override val message: String

    fun toJSON(): Map<String, Any?> = mapOf(
        "code" to toCode(this),
        "message" to (this.message ?: ""),
        "platform" to "android",
    )

    class ProductNotFound(val productId: String) : OpenIapError() {
        val CODE = ErrorCode.SkuNotFound.rawValue
        override val code = CODE
        override val message = MESSAGE

        companion object {
            val CODE = ErrorCode.SkuNotFound.rawValue
            const val MESSAGE = "Product not found"
        }
    }

    object PurchaseFailed : OpenIapError() {
        val CODE = ErrorCode.PurchaseError.rawValue
        override val code = CODE
        override val message = MESSAGE

        const val MESSAGE = "Purchase failed"
    }

    object PurchaseCancelled : OpenIapError() {
        val CODE = ErrorCode.UserCancelled.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Purchase was cancelled by the user"
    }

    object PurchaseDeferred : OpenIapError() {
        val CODE = ErrorCode.DeferredPayment.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Purchase was deferred"
    }

    object PaymentNotAllowed : OpenIapError() {
        val CODE = ErrorCode.UserError.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Payment not allowed"
    }

    object BillingError : OpenIapError() {
        val CODE = ErrorCode.ServiceError.rawValue
        override val code = CODE
        override val message = MESSAGE

        const val MESSAGE = "Billing error"
    }

    /**
     * @deprecated Use [InvalidPurchaseVerification] instead
     */
    @Deprecated("Use InvalidPurchaseVerification instead", ReplaceWith("InvalidPurchaseVerification"))
    object InvalidReceipt : OpenIapError() {
        val CODE = ErrorCode.PurchaseVerificationFailed.rawValue
        override val code = CODE
        override val message = MESSAGE

        const val MESSAGE = "Purchase verification failed"
    }

    /**
     * Purchase verification failed (general error without specific provider message)
     */
    object InvalidPurchaseVerification : OpenIapError() {
        val CODE = ErrorCode.PurchaseVerificationFailed.rawValue
        override val code = CODE
        override val message = MESSAGE

        const val MESSAGE = "Purchase verification failed"
    }

    /**
     * Purchase verification failed with a specific error from the verification provider (e.g., IAPKit)
     */
    class PurchaseVerificationFailed(val providerError: String) : OpenIapError() {
        override val code = ErrorCode.PurchaseVerificationFailed.rawValue
        override val message = "Purchase verification failed: $providerError"

        companion object {
            val CODE = ErrorCode.PurchaseVerificationFailed.rawValue
        }
    }

    object NetworkError : OpenIapError() {
        val CODE = ErrorCode.NetworkError.rawValue
        const val MESSAGE = "Network connection error"
        override val code: String = CODE
        override val message: String = MESSAGE
    }

    object VerificationFailed : OpenIapError() {
        val CODE = ErrorCode.TransactionValidationFailed.rawValue
        override val code = CODE
        override val message = MESSAGE

        const val MESSAGE = "Verification failed"
    }

    object RestoreFailed : OpenIapError() {
        val CODE = ErrorCode.SyncError.rawValue
        override val code = CODE
        override val message = MESSAGE

        const val MESSAGE = "Restore failed"
    }

    object UnknownError : OpenIapError() {
        val CODE = ErrorCode.Unknown.rawValue
        override val code = CODE
        override val message = MESSAGE

        const val MESSAGE = "Unknown error"
    }

    object NotPrepared : OpenIapError() {
        const val CODE = "not-prepared"
        const val MESSAGE = "Billing client not ready"
        override val code: String = CODE
        override val message: String = MESSAGE
    }

    object InitConnection : OpenIapError() {
        val CODE = ErrorCode.InitConnection.rawValue
        override val code = CODE
        override val message = MESSAGE

        const val MESSAGE = "Failed to initialize billing connection"
    }

    object QueryProduct : OpenIapError() {
        val CODE = ErrorCode.QueryProduct.rawValue
        override val code = CODE
        override val message = MESSAGE

        const val MESSAGE = "Failed to query product"
    }

    object EmptySkuList : OpenIapError() {
        const val CODE = "empty-sku-list"
        const val MESSAGE = "SKU list cannot be empty"
        override val code: String = CODE
        override val message: String = MESSAGE
    }

    class SkuNotFound(val sku: String) : OpenIapError() {
        val CODE = ErrorCode.SkuNotFound.rawValue
        override val code = CODE
        override val message = MESSAGE

        companion object {
            val CODE = ErrorCode.SkuNotFound.rawValue
            const val MESSAGE = "SKU not found"
        }
    }

    object SkuOfferMismatch : OpenIapError() {
        const val CODE = "sku-offer-mismatch"
        const val MESSAGE = "SKU and offer token count mismatch"
        override val code: String = CODE
        override val message: String = MESSAGE
    }

    object MissingCurrentActivity : OpenIapError() {
        val CODE = ErrorCode.ActivityUnavailable.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Current activity is not available"
    }

    object UserCancelled : OpenIapError() {
        val CODE = ErrorCode.UserCancelled.rawValue
        const val MESSAGE = "User cancelled the operation"
        override val code: String = CODE
        override val message: String = MESSAGE
    }

    object ItemAlreadyOwned : OpenIapError() {
        val CODE = ErrorCode.AlreadyOwned.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Item is already owned"
    }

    object ItemNotOwned : OpenIapError() {
        val CODE = ErrorCode.ItemNotOwned.rawValue
        const val MESSAGE = "Item is not owned"
        override val code: String = CODE
        override val message: String = MESSAGE
    }

    object ServiceUnavailable : OpenIapError() {
        val CODE = ErrorCode.ServiceError.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Billing service is unavailable"
    }

    object BillingUnavailable : OpenIapError() {
        val CODE = ErrorCode.BillingUnavailable.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Billing API version is not supported"
    }

    object ItemUnavailable : OpenIapError() {
        val CODE = ErrorCode.ItemUnavailable.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Requested product is not available for purchase"
    }

    object DeveloperError : OpenIapError() {
        val CODE = ErrorCode.DeveloperError.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Invalid arguments provided to the API"
    }

    object FeatureNotSupported : OpenIapError() {
        val CODE = ErrorCode.FeatureNotSupported.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Requested feature is not supported by Play Store"
    }

    object ServiceDisconnected : OpenIapError() {
        val CODE = ErrorCode.ServiceDisconnected.rawValue
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "Play Store service is not connected"
    }

    object ServiceTimeout : OpenIapError() {
        const val CODE = "service-timeout"
        override val code: String = CODE
        override val message: String = MESSAGE

        const val MESSAGE = "The request has reached the maximum timeout before billing service responds"
    }

    class AlternativeBillingUnavailable(val details: String) : OpenIapError() {
        val CODE = ErrorCode.BillingUnavailable.rawValue
        override val code = CODE
        override val message = details

        companion object {
            val CODE = ErrorCode.BillingUnavailable.rawValue
        }
    }

    companion object {
        private val defaultMessages: Map<String, String> by lazy {
            mapOf(
                ProductNotFound.CODE to ProductNotFound.MESSAGE,
                PurchaseFailed.CODE to PurchaseFailed.MESSAGE,
                PurchaseCancelled.CODE to PurchaseCancelled.MESSAGE,
                PurchaseDeferred.CODE to PurchaseDeferred.MESSAGE,
                NetworkError.CODE to NetworkError.MESSAGE,
                UnknownError.CODE to UnknownError.MESSAGE,
                NotPrepared.CODE to NotPrepared.MESSAGE,
                InitConnection.CODE to InitConnection.MESSAGE,
                QueryProduct.CODE to QueryProduct.MESSAGE,
                EmptySkuList.CODE to EmptySkuList.MESSAGE,
                SkuNotFound.CODE to SkuNotFound.MESSAGE,
                SkuOfferMismatch.CODE to SkuOfferMismatch.MESSAGE,
                UserCancelled.CODE to UserCancelled.MESSAGE,
                ItemAlreadyOwned.CODE to ItemAlreadyOwned.MESSAGE,
                ItemNotOwned.CODE to ItemNotOwned.MESSAGE,
                ServiceUnavailable.CODE to ServiceUnavailable.MESSAGE,
                BillingUnavailable.CODE to BillingUnavailable.MESSAGE,
                ItemUnavailable.CODE to ItemUnavailable.MESSAGE,
                DeveloperError.CODE to DeveloperError.MESSAGE,
                FeatureNotSupported.CODE to FeatureNotSupported.MESSAGE,
                ServiceDisconnected.CODE to ServiceDisconnected.MESSAGE,
                ServiceTimeout.CODE to ServiceTimeout.MESSAGE,
                PaymentNotAllowed.CODE to PaymentNotAllowed.MESSAGE,
                BillingError.CODE to BillingError.MESSAGE,
                InvalidPurchaseVerification.CODE to InvalidPurchaseVerification.MESSAGE,
                VerificationFailed.CODE to VerificationFailed.MESSAGE,
                RestoreFailed.CODE to RestoreFailed.MESSAGE,
                MissingCurrentActivity.CODE to MissingCurrentActivity.MESSAGE
            )
        }

        fun toCode(error: OpenIapError): String = error.code

        fun defaultMessage(code: String): String =
            defaultMessages[code] ?: "Unknown error occurred"

        fun getAllErrorCodes(): Map<String, String> = defaultMessages
    }

}

File: packages/apple/Sources/Models/OpenIapError.swift
import Foundation
import StoreKit

public extension PurchaseError {
    // MARK: - Default Messages

    static func defaultMessage(for code: ErrorCode) -> String {
        switch code {
        case .unknown: return "Unknown error occurred"
        case .userCancelled: return "User cancelled the purchase flow"
        case .userError: return "User action error"
        case .itemUnavailable: return "Item unavailable"
        case .remoteError: return "Remote service error"
        case .networkError: return "Network connection error"
        case .serviceError: return "Store service error"
        case .purchaseVerificationFailed: return "Purchase verification failed"
        case .purchaseVerificationFinished: return "Transaction already finished"
        case .purchaseVerificationFinishFailed: return "Transaction finish failed"
        case .notPrepared: return "Billing is not prepared"
        case .notEnded: return "Billing connection not ended"
        case .alreadyOwned: return "Item already owned"
        case .developerError: return "Developer configuration error"
        case .billingResponseJsonParseError: return "Failed to parse billing response"
        case .deferredPayment: return "Payment was deferred (pending approval)"
        case .interrupted: return "Purchase flow interrupted"
        case .iapNotAvailable: return "In-app purchases not available on this device"
        case .purchaseError: return "Purchase error"
        case .syncError: return "Sync error"
        case .transactionValidationFailed: return "Transaction validation failed"
        case .activityUnavailable: return "Required activity is unavailable"
        case .alreadyPrepared: return "Billing already prepared"
        case .pending: return "Transaction pending"
        case .connectionClosed: return "Connection closed"
        case .initConnection: return "Failed to initialize billing connection"
        case .serviceDisconnected: return "Billing service disconnected"
        case .serviceTimeout: return "Billing service request timed out"
        case .queryProduct: return "Failed to query product"
        case .skuNotFound: return "SKU not found"
        case .skuOfferMismatch: return "SKU offer mismatch"
        case .itemNotOwned: return "Item not owned"
        case .billingUnavailable: return "Billing unavailable"
        case .featureNotSupported: return "Feature not supported on this platform"
        case .emptySkuList: return "Empty SKU list provided"
        case .duplicatePurchase: return "Duplicate purchase update detected"
        }
    }

    static func defaultMessage(for rawCode: String) -> String {
        if let parsed = ErrorCode(rawValue: rawCode) {
            return defaultMessage(for: parsed)
        }
        return "Unknown error occurred"
    }

    static func make(
        code: ErrorCode,
        productId: String? = nil,
        message: String? = nil
    ) -> PurchaseError {
        make(
            code: code,
            productId: productId,
            message: message,
            debugMessage: nil
        )
    }

    static func make(
        code: ErrorCode,
        productId: String? = nil,
        message: String? = nil,
        debugMessage: String?
    ) -> PurchaseError {
        PurchaseError(
            code: code,
            debugMessage: debugMessage,
            message: message ?? defaultMessage(for: code),
            productId: productId
        )
    }

    // MARK: - Convenience Constructors

    static func make(
        code: String,
        productId: String? = nil,
        message: String? = nil
    ) -> PurchaseError {
        make(
            code: code,
            productId: productId,
            message: message,
            debugMessage: nil
        )
    }

    static func make(
        code: String,
        productId: String? = nil,
        message: String? = nil,
        debugMessage: String?
    ) -> PurchaseError {
        let resolved = ErrorCode(rawValue: code) ?? .unknown
        return make(
            code: resolved,
            productId: productId,
            message: message,
            debugMessage: debugMessage
        )
    }

    static func emptySkuList(message: String? = nil) -> PurchaseError {
        make(code: .emptySkuList, message: message)
    }

    static func purchaseError(message: String? = nil, productId: String? = nil) -> PurchaseError {
        make(code: .purchaseError, productId: productId, message: message)
    }

    /// Returns the canonical set of error codes mapped to their default messages.
    static func errorCodeTable() -> [String: String] {
        ErrorCode.allCases.reduce(into: [String: String]()) { result, code in
            result[code.rawValue] = defaultMessage(for: code)
        }
    }

    /// Wraps any error into a `PurchaseError`, preserving existing instances.
    /// Automatically maps StoreKit errors to appropriate PurchaseError codes.
    static func wrap(
        _ error: Error,
        fallback: ErrorCode = .purchaseError,
        productId: String? = nil
    ) -> PurchaseError {
        func isStoreKitCancellation(_ error: Error) -> Bool {
            if let storeKitError = error as? StoreKitError {
                switch storeKitError {
                case .userCancelled:
                    return true
                case .systemError(let underlyingError):
                    return isStoreKitCancellation(underlyingError)
                default:
                    return false
                }
            }

            if let skError = error as? SKError {
                return skError.code == .paymentCancelled
            }

            let nsError = error as NSError
            return nsError.domain == SKError.errorDomain &&
                nsError.code == SKError.Code.paymentCancelled.rawValue
        }

        // If already a PurchaseError, return as-is
        if let purchaseError = error as? PurchaseError {
            return purchaseError
        }

        if isStoreKitCancellation(error) {
            return make(
                code: .userCancelled,
                productId: productId,
                message: error.localizedDescription,
                debugMessage: error.localizedDescription
            )
        }

        // Map StoreKit 2 errors to PurchaseError
        if let storeKitError = error as? StoreKitError {
            let mappedCode: ErrorCode
            switch storeKitError {
            case .userCancelled:
                mappedCode = .userCancelled
            case .networkError:
                mappedCode = .networkError
            case .notAvailableInStorefront:
                mappedCode = .itemUnavailable
            case .notEntitled:
                mappedCode = .itemNotOwned
            case .systemError(let underlyingError):
                if let skErrorCode = skErrorCode(from: underlyingError) {
                    mappedCode = errorCode(for: skErrorCode) ?? .serviceError
                } else {
                    mappedCode = .serviceError
                }
            case .unknown:
                mappedCode = .unknown
            case .unsupported:
                mappedCode = .featureNotSupported
            @unknown default:
                mappedCode = fallback
            }
            return make(
                code: mappedCode,
                productId: productId,
                message: error.localizedDescription,
                debugMessage: error.localizedDescription
            )
        }

        // StoreKit 1 conditions reach `wrap` through promoted purchases,
        // offer-code redemption, and the legacy payment queue.
        if let skErrorCode = skErrorCode(from: error) {
            return make(
                code: errorCode(for: skErrorCode) ?? fallback,
                productId: productId,
                message: error.localizedDescription,
                debugMessage: error.localizedDescription
            )
        }

        // Fallback for other error types
        return make(
            code: fallback,
            productId: productId,
            message: error.localizedDescription,
            debugMessage: error.localizedDescription
        )
    }

    /// Accepts a typed `SKError` or a bridged `NSError` in the StoreKit domain.
    static func skErrorCode(from error: Error) -> SKError.Code? {
        if let skError = error as? SKError {
            return skError.code
        }
        let nsError = error as NSError
        guard nsError.domain == SKError.errorDomain else { return nil }
        return SKError.Code(rawValue: nsError.code)
    }

    /// Normative StoreKit 1 error mapping. Returns `nil` when no OpenIAP code
    /// faithfully represents the condition, so the caller's `fallback` applies
    /// rather than a fabricated mapping.
    ///
    /// `.alreadyOwned`, `.billingUnavailable`, `.serviceDisconnected`, and
    /// `.serviceTimeout` are deliberately unreachable here — StoreKit has no
    /// equivalent condition. See `specs/client/src/capability-matrix.mjs`.
    static func errorCode(for code: SKError.Code) -> ErrorCode? {
        switch code {
        case .paymentCancelled:
            return .userCancelled
        case .paymentNotAllowed:
            // Parental controls or MDM restriction, not a transient failure.
            return .iapNotAvailable
        case .storeProductNotAvailable:
            return .itemUnavailable
        case .clientInvalid, .paymentInvalid:
            return .developerError
        case .cloudServiceNetworkConnectionFailed:
            return .networkError
        case .cloudServicePermissionDenied, .cloudServiceRevoked, .privacyAcknowledgementRequired:
            return .serviceError
        case .invalidOfferIdentifier, .invalidOfferPrice, .missingOfferParams:
            return .skuOfferMismatch
        case .invalidSignature, .unauthorizedRequestData:
            return .transactionValidationFailed
        case .unknown:
            return .unknown
        default:
            return nil
        }
    }
}
